In [1]:
import numpy as np
from tensorflow import keras
from tensorflow.keras import layers
import zipfile
import os
import pandas as pd 
from sklearn.model_selection import train_test_split
import keras.preprocessing as kp

# Mnist

## Preprocessing

In [2]:
num_classes = 10
input_shape = (28, 28, 1)

(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()

x_train = x_train.astype("float32") / 255
x_test = x_test.astype("float32") / 255
x_train = np.expand_dims(x_train, -1)
x_test = np.expand_dims(x_test, -1)

y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)

11490434/11490434 [==============================] - 10s 1us/step


## Model 1
Prosty model - splot(32), splot(32), warstwa gęsta(10)

In [3]:
model1 = keras.Sequential(
    [
        keras.Input(shape=input_shape),
        layers.Conv2D(32, kernel_size=(3, 3), activation="relu"),
        layers.MaxPooling2D(pool_size=(2, 2)),
        layers.Conv2D(64, kernel_size=(3, 3), activation="relu"),
        layers.MaxPooling2D(pool_size=(2, 2)),
        layers.Flatten(),
        layers.Dropout(0.5),
        layers.Dense(num_classes, activation="softmax"),
    ]
)

model1.summary()


Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 26, 26, 32)        320       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 13, 13, 32)       0         
 )                                                               
                                                                 
 conv2d_1 (Conv2D)           (None, 11, 11, 64)        18496     
                                                                 
 max_pooling2d_1 (MaxPooling  (None, 5, 5, 64)         0         
 2D)                                                             
                                                                 
 flatten (Flatten)           (None, 1600)              0         
                                                                 
 dropout (Dropout)           (None, 1600)              0

In [4]:
batch_size = 128
epochs = 15

model1.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"])

model1.fit(x_train, y_train, batch_size=batch_size, epochs=epochs, validation_split=0.1)

Epoch 1/15
422/422 [==============================] - 29s 68ms/step - loss: 0.3662 - accuracy: 0.8878 - val_loss: 0.0796 - val_accuracy: 0.9787
Epoch 2/15
422/422 [==============================] - 29s 69ms/step - loss: 0.1126 - accuracy: 0.9653 - val_loss: 0.0635 - val_accuracy: 0.9813
Epoch 3/15
422/422 [==============================] - 30s 71ms/step - loss: 0.0852 - accuracy: 0.9736 - val_loss: 0.0522 - val_accuracy: 0.9850
Epoch 4/15
422/422 [==============================] - 30s 71ms/step - loss: 0.0714 - accuracy: 0.9781 - val_loss: 0.0422 - val_accuracy: 0.9883
Epoch 5/15
422/422 [==============================] - 35s 83ms/step - loss: 0.0636 - accuracy: 0.9799 - val_loss: 0.0413 - val_accuracy: 0.9898
Epoch 6/15
422/422 [==============================] - 29s 70ms/step - loss: 0.0582 - accuracy: 0.9816 - val_loss: 0.0392 - val_accuracy: 0.9885
Epoch 7/15
422/422 [==============================] - 34s 82ms/step - loss: 0.0534 - accuracy: 0.9831 - val_loss: 0.0371 - val_accuracy:

In [8]:
score = model1.evaluate(x_test, y_test, verbose=0)
print("Test loss:", score[0])
print("Test accuracy:", score[1])

Test loss: 0.02691200003027916
Test accuracy: 0.9905999898910522


Skueczność - 99,06%

## Model 2
Dorzucam Dropout(0.25)

In [6]:
model2 = keras.Sequential(
    [
        keras.Input(shape=input_shape),
        layers.Conv2D(32, kernel_size=(3, 3), activation="relu"),
        layers.MaxPooling2D(pool_size=(2, 2)),
        layers.Dropout(0.25),
        layers.Conv2D(64, kernel_size=(3, 3), activation="relu"),
        layers.MaxPooling2D(pool_size=(2, 2)),
        layers.Dropout(0.25),
        layers.Flatten(),
        layers.Dense(num_classes, activation="softmax"),
    ]
)

model2.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_2 (Conv2D)           (None, 26, 26, 32)        320       
                                                                 
 max_pooling2d_2 (MaxPooling  (None, 13, 13, 32)       0         
 2D)                                                             
                                                                 
 dropout_1 (Dropout)         (None, 13, 13, 32)        0         
                                                                 
 conv2d_3 (Conv2D)           (None, 11, 11, 64)        18496     
                                                                 
 max_pooling2d_3 (MaxPooling  (None, 5, 5, 64)         0         
 2D)                                                             
                                                                 
 dropout_2 (Dropout)         (None, 5, 5, 64)         

In [7]:
batch_size = 128
epochs = 15

model2.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"])

model2.fit(x_train, y_train, batch_size=batch_size, epochs=epochs, validation_split=0.1)

Epoch 1/15
422/422 [==============================] - 32s 74ms/step - loss: 0.4506 - accuracy: 0.8574 - val_loss: 0.0981 - val_accuracy: 0.9750
Epoch 2/15
422/422 [==============================] - 31s 74ms/step - loss: 0.1500 - accuracy: 0.9545 - val_loss: 0.0695 - val_accuracy: 0.9807
Epoch 3/15
422/422 [==============================] - 35s 84ms/step - loss: 0.1172 - accuracy: 0.9641 - val_loss: 0.0537 - val_accuracy: 0.9855
Epoch 4/15
422/422 [==============================] - 31s 75ms/step - loss: 0.0967 - accuracy: 0.9697 - val_loss: 0.0462 - val_accuracy: 0.9882
Epoch 5/15
422/422 [==============================] - 34s 80ms/step - loss: 0.0877 - accuracy: 0.9728 - val_loss: 0.0409 - val_accuracy: 0.9885
Epoch 6/15
422/422 [==============================] - 34s 81ms/step - loss: 0.0835 - accuracy: 0.9740 - val_loss: 0.0375 - val_accuracy: 0.9898
Epoch 7/15
422/422 [==============================] - 34s 81ms/step - loss: 0.0764 - accuracy: 0.9764 - val_loss: 0.0369 - val_accuracy:

In [9]:
score = model2.evaluate(x_test, y_test, verbose=0)
print("Test loss:", score[0])
print("Test accuracy:", score[1])

Test loss: 0.026326607912778854
Test accuracy: 0.9907000064849854


Skuteczność - 99,07%

## Model 3
Dorzucam warstwy gęste (128, 30)

In [10]:
model3 = keras.Sequential(
    [
        keras.Input(shape=input_shape),
        layers.Conv2D(32, kernel_size=(3, 3), activation="relu"),
        layers.MaxPooling2D(pool_size=(2, 2)),
        layers.Dropout(0.25),
        layers.Conv2D(64, kernel_size=(3, 3), activation="relu"),
        layers.MaxPooling2D(pool_size=(2, 2)),
        layers.Dropout(0.25),
        layers.Flatten(),
        layers.Dense(128),
        layers.Dense(30),
        layers.Dense(num_classes, activation="softmax"),
    ]
)

model3.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_4 (Conv2D)           (None, 26, 26, 32)        320       
                                                                 
 max_pooling2d_4 (MaxPooling  (None, 13, 13, 32)       0         
 2D)                                                             
                                                                 
 dropout_4 (Dropout)         (None, 13, 13, 32)        0         
                                                                 
 conv2d_5 (Conv2D)           (None, 11, 11, 64)        18496     
                                                                 
 max_pooling2d_5 (MaxPooling  (None, 5, 5, 64)         0         
 2D)                                                             
                                                                 
 dropout_5 (Dropout)         (None, 5, 5, 64)         

In [14]:
batch_size = 128
epochs = 15

model3.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"])

model3.fit(x_train, y_train, batch_size=batch_size, epochs=epochs, validation_split=0.1)

Epoch 1/15
422/422 [==============================] - 34s 78ms/step - loss: 0.2489 - accuracy: 0.9226 - val_loss: 0.0646 - val_accuracy: 0.9812
Epoch 2/15
422/422 [==============================] - 35s 83ms/step - loss: 0.0840 - accuracy: 0.9734 - val_loss: 0.0451 - val_accuracy: 0.9860
Epoch 3/15
422/422 [==============================] - 35s 83ms/step - loss: 0.0641 - accuracy: 0.9803 - val_loss: 0.0398 - val_accuracy: 0.9888
Epoch 4/15
422/422 [==============================] - 35s 83ms/step - loss: 0.0526 - accuracy: 0.9832 - val_loss: 0.0391 - val_accuracy: 0.9888
Epoch 5/15
422/422 [==============================] - 35s 83ms/step - loss: 0.0463 - accuracy: 0.9848 - val_loss: 0.0367 - val_accuracy: 0.9897
Epoch 6/15
422/422 [==============================] - 36s 84ms/step - loss: 0.0419 - accuracy: 0.9864 - val_loss: 0.0356 - val_accuracy: 0.9887
Epoch 7/15
422/422 [==============================] - 35s 83ms/step - loss: 0.0409 - accuracy: 0.9870 - val_loss: 0.0337 - val_accuracy:

In [15]:
score = model3.evaluate(x_test, y_test, verbose=0)
print("Test loss:", score[0])
print("Test accuracy:", score[1])

Test loss: 0.02585824951529503
Test accuracy: 0.9919000267982483


Skuteczność: 99,19%

# Dogs vs. Cats

## Preprocessing

In [2]:
zip_path='train.zip'
save_path = 'DogsCats'
file=zipfile.ZipFile(zip_path)
file.extractall(save_path)
file.close()

In [3]:
zip_path='test1.zip'
save_path = 'DogsCats'
file=zipfile.ZipFile(zip_path)
file.extractall(save_path)
file.close()

In [4]:
filenames = os.listdir("DogsCats/train")
categories = []
for image in filenames :
    category = image.split('.')[0]
    if category == 'dog' :
        categories.append(1)
    else :    
        categories.append(0)

img_height, img_width = 128, 128
batch_size = 32

In [5]:
df = pd.DataFrame({'filename' : filenames, 'category' : categories})
df['category'] = df['category'].map({0 : 'cat', 1 : 'dog'})

In [6]:
train_df, valid_df = train_test_split(df, test_size=0.2, random_state=42)
train_df = train_df.reset_index(drop=True)
valid_df = valid_df.reset_index(drop=True)

In [7]:
total_train = train_df.shape[0]
total_valid = valid_df.shape[0]

In [8]:
train_data = kp.image.ImageDataGenerator(rotation_range=20,
                                                         rescale = 1./255,
                                                         horizontal_flip=True)
train_generator = train_data.flow_from_dataframe(train_df,
                                                'DogsCats/train/',
                                                x_col = 'filename',
                                                y_col = 'category',
                                                target_size=[128,128],
                                                class_mode='categorical',
                                                batch_size=batch_size)

Found 20000 validated image filenames belonging to 2 classes.


In [9]:
valid_data = keras.preprocessing.image.ImageDataGenerator(rescale = 1./255)
                                                         
                                                         
valid_generator = valid_data.flow_from_dataframe(valid_df,
                                                'DogsCats/train/',
                                                x_col = 'filename',
                                                y_col = 'category',
                                                target_size=[128,128],
                                                class_mode='categorical',
                                                batch_size=batch_size)

Found 5000 validated image filenames belonging to 2 classes.


## Model 4

In [10]:
model4 = keras.Sequential()
model4.add(layers.Conv2D(64, kernel_size=(3, 3), activation='relu', input_shape=(img_height, img_width, 3)))
model4.add(layers.MaxPooling2D(pool_size=(2, 2)))
model4.add(layers.Dropout(0.25))
model4.add(layers.Conv2D(128, kernel_size=(3, 3), activation='relu'))
model4.add(layers.MaxPooling2D(pool_size=(2, 2)))
model4.add(layers.Dropout(0.25))
model4.add(layers.Flatten())
model4.add(layers.Dense(512, activation='relu'))
model4.add(layers.Dense(2, activation = 'softmax'))
 
model4.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
 
model4.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 126, 126, 64)      1792      
                                                                 
 max_pooling2d (MaxPooling2D  (None, 63, 63, 64)       0         
 )                                                               
                                                                 
 dropout (Dropout)           (None, 63, 63, 64)        0         
                                                                 
 conv2d_1 (Conv2D)           (None, 61, 61, 128)       73856     
                                                                 
 max_pooling2d_1 (MaxPooling  (None, 30, 30, 128)      0         
 2D)                                                             
                                                                 
 dropout_1 (Dropout)         (None, 30, 30, 128)       0

In [10]:
earlystop = keras.callbacks.EarlyStopping(patience=5)
learning_rate_reduction = keras.callbacks.ReduceLROnPlateau(monitor='val_accuracy', 
                                            patience=2, 
                                            verbose=1, 
                                            factor=0.5, 
                                            min_lr=0.00001)
callbacks = [earlystop, learning_rate_reduction]

In [12]:
epochs = 15
model4.fit_generator(
    train_generator, 
    epochs=epochs,
    validation_data=valid_generator,
    validation_steps=total_valid//batch_size,
    steps_per_epoch=total_train//batch_size,
    callbacks=callbacks)

<ipython-input-12-3298df0a7386>:2: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.
  model4.fit_generator(


Epoch 1/15
625/625 [==============================] - 1057s 2s/step - loss: 0.7536 - accuracy: 0.5552 - val_loss: 0.6733 - val_accuracy: 0.5861 - lr: 0.0010
Epoch 2/15
625/625 [==============================] - 988s 2s/step - loss: 0.6413 - accuracy: 0.6303 - val_loss: 0.6111 - val_accuracy: 0.6707 - lr: 0.0010
Epoch 3/15
625/625 [==============================] - 985s 2s/step - loss: 0.5999 - accuracy: 0.6811 - val_loss: 0.5559 - val_accuracy: 0.7121 - lr: 0.0010
Epoch 4/15
625/625 [==============================] - 979s 2s/step - loss: 0.5525 - accuracy: 0.7174 - val_loss: 0.5222 - val_accuracy: 0.7480 - lr: 0.0010
Epoch 5/15
625/625 [==============================] - 978s 2s/step - loss: 0.5118 - accuracy: 0.7463 - val_loss: 0.4836 - val_accuracy: 0.7702 - lr: 0.0010
Epoch 6/15
625/625 [==============================] - 978s 2s/step - loss: 0.4788 - accuracy: 0.7685 - val_loss: 0.4507 - val_accuracy: 0.7871 - lr: 0.0010
Epoch 7/15
625/625 [==============================] - 983s 2s/s

loss: 0.3311 - accuracy: 0.8544 - val_loss: 0.4204 - val_accuracy: 0.8111

## Model 5

In [13]:
model5 = keras.Sequential()
model5.add(layers.Conv2D(32, kernel_size=(3, 3), activation='relu', input_shape=(img_height, img_width, 3)))
model5.add(layers.MaxPooling2D(pool_size=(2, 2)))
model5.add(layers.Dropout(0.25))
model5.add(layers.Conv2D(64, kernel_size=(3, 3), activation='relu'))
model5.add(layers.MaxPooling2D(pool_size=(2, 2)))
model5.add(layers.Dropout(0.25))
model5.add(layers.Flatten())
model5.add(layers.Dense(2, activation = 'softmax'))
 
model5.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
 
model5.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_2 (Conv2D)           (None, 126, 126, 32)      896       
                                                                 
 max_pooling2d_2 (MaxPooling  (None, 63, 63, 32)       0         
 2D)                                                             
                                                                 
 dropout_2 (Dropout)         (None, 63, 63, 32)        0         
                                                                 
 conv2d_3 (Conv2D)           (None, 61, 61, 64)        18496     
                                                                 
 max_pooling2d_3 (MaxPooling  (None, 30, 30, 64)       0         
 2D)                                                             
                                                                 
 dropout_3 (Dropout)         (None, 30, 30, 64)       

In [14]:
epochs = 30
model5.fit_generator(
    train_generator, 
    epochs=epochs,
    validation_data=valid_generator,
    validation_steps=total_valid//batch_size,
    steps_per_epoch=total_train//batch_size,
    callbacks=callbacks)

<ipython-input-14-82596d7ab951>:2: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.
  model5.fit_generator(


Epoch 1/30
625/625 [==============================] - 379s 605ms/step - loss: 0.6729 - accuracy: 0.5867 - val_loss: 0.6304 - val_accuracy: 0.6456 - lr: 0.0010
Epoch 2/30
625/625 [==============================] - 382s 610ms/step - loss: 0.5647 - accuracy: 0.7110 - val_loss: 0.5385 - val_accuracy: 0.7348 - lr: 0.0010
Epoch 3/30
625/625 [==============================] - 379s 606ms/step - loss: 0.5158 - accuracy: 0.7474 - val_loss: 0.4737 - val_accuracy: 0.7728 - lr: 0.0010
Epoch 4/30
625/625 [==============================] - 378s 604ms/step - loss: 0.4876 - accuracy: 0.7653 - val_loss: 0.4628 - val_accuracy: 0.7899 - lr: 0.0010
Epoch 5/30
625/625 [==============================] - 380s 608ms/step - loss: 0.4793 - accuracy: 0.7736 - val_loss: 0.4531 - val_accuracy: 0.7909 - lr: 0.0010
Epoch 6/30
625/625 [==============================] - 378s 605ms/step - loss: 0.4590 - accuracy: 0.7880 - val_loss: 0.4368 - val_accuracy: 0.7921 - lr: 0.0010
Epoch 7/30
625/625 [==========================

loss: 0.3337 - accuracy: 0.8554 - val_loss: 0.3510 - val_accuracy: 0.8415

## Model 6

In [11]:
model6 = keras.Sequential()
model6.add(layers.Conv2D(32, kernel_size=(3, 3), activation='relu', input_shape=(img_height, img_width, 3)))
model6.add(layers.Conv2D(32, kernel_size=(3, 3), activation='relu', input_shape=(img_height, img_width, 3)))
model6.add(layers.MaxPooling2D(pool_size=(2, 2)))
model6.add(layers.Dropout(0.25))
model6.add(layers.Conv2D(64, kernel_size=(3, 3), activation='relu'))
model6.add(layers.Conv2D(64, kernel_size=(3, 3), activation='relu'))
model6.add(layers.MaxPooling2D(pool_size=(2, 2)))
model6.add(layers.Dropout(0.25))
model6.add(layers.Flatten())
model6.add(layers.Dense(128, activation = 'softmax'))
model6.add(layers.Dense(2, activation = 'softmax'))
 
model6.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
 
model6.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 126, 126, 32)      896       
                                                                 
 conv2d_1 (Conv2D)           (None, 124, 124, 32)      9248      
                                                                 
 max_pooling2d (MaxPooling2D  (None, 62, 62, 32)       0         
 )                                                               
                                                                 
 dropout (Dropout)           (None, 62, 62, 32)        0         
                                                                 
 conv2d_2 (Conv2D)           (None, 60, 60, 64)        18496     
                                                                 
 conv2d_3 (Conv2D)           (None, 58, 58, 64)        36928     
                                                        

In [12]:
epochs = 30
model6.fit_generator(
    train_generator, 
    epochs=epochs,
    validation_data=valid_generator,
    validation_steps=total_valid//batch_size,
    steps_per_epoch=total_train//batch_size,
    callbacks=callbacks)

<ipython-input-12-003318f83f9e>:2: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.
  model6.fit_generator(


Epoch 1/30
625/625 [==============================] - 1065s 2s/step - loss: 0.6934 - accuracy: 0.4999 - val_loss: 0.6931 - val_accuracy: 0.5034 - lr: 0.0010
Epoch 2/30
625/625 [==============================] - 952s 2s/step - loss: 0.6932 - accuracy: 0.5041 - val_loss: 0.6932 - val_accuracy: 0.4972 - lr: 0.0010
Epoch 3/30
625/625 [==============================] - ETA: 0s - loss: 0.6933 - accuracy: 0.4956
Epoch 3: ReduceLROnPlateau reducing learning rate to 0.0005000000237487257.
625/625 [==============================] - 942s 2s/step - loss: 0.6933 - accuracy: 0.4956 - val_loss: 0.6932 - val_accuracy: 0.4976 - lr: 0.0010
Epoch 4/30
625/625 [==============================] - 973s 2s/step - loss: 0.6932 - accuracy: 0.4985 - val_loss: 0.6932 - val_accuracy: 0.4964 - lr: 5.0000e-04
Epoch 5/30
625/625 [==============================] - ETA: 0s - loss: 0.6932 - accuracy: 0.4958
Epoch 5: ReduceLROnPlateau reducing learning rate to 0.0002500000118743628.
625/625 [=============================

loss: 0.6932 - accuracy: 0.4970 - val_loss: 0.6931 - val_accuracy: 0.5030 - lr: 2.5000e-04

## Model 7

In [13]:
model7 = keras.Sequential()
model7.add(layers.Conv2D(64, kernel_size=(3, 3), activation='relu', input_shape=(img_height, img_width, 3)))
model7.add(layers.AveragePooling2D(pool_size=(2, 2)))
model7.add(layers.BatchNormalization())
model7.add(layers.Conv2D(64, kernel_size=(3, 3), activation='relu'))
model7.add(layers.AveragePooling2D(pool_size=(2, 2)))
model7.add(layers.BatchNormalization())
model7.add(layers.Flatten())
model7.add(layers.Dense(2, activation = 'softmax'))
 
model7.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
 
model7.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_4 (Conv2D)           (None, 126, 126, 64)      1792      
                                                                 
 average_pooling2d (AverageP  (None, 63, 63, 64)       0         
 ooling2D)                                                       
                                                                 
 batch_normalization (BatchN  (None, 63, 63, 64)       256       
 ormalization)                                                   
                                                                 
 conv2d_5 (Conv2D)           (None, 61, 61, 64)        36928     
                                                                 
 average_pooling2d_1 (Averag  (None, 30, 30, 64)       0         
 ePooling2D)                                                     
                                                      

In [14]:
epochs = 30
model7.fit_generator(
    train_generator, 
    epochs=epochs,
    validation_data=valid_generator,
    validation_steps=total_valid//batch_size,
    steps_per_epoch=total_train//batch_size,
    callbacks=callbacks)

<ipython-input-14-bb1c34f6d3dc>:2: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.
  model7.fit_generator(


Epoch 1/30
625/625 [==============================] - 569s 908ms/step - loss: 3.5706 - accuracy: 0.5907 - val_loss: 2.2989 - val_accuracy: 0.5747 - lr: 0.0010
Epoch 2/30
625/625 [==============================] - 542s 867ms/step - loss: 1.3133 - accuracy: 0.6445 - val_loss: 0.9691 - val_accuracy: 0.6753 - lr: 0.0010
Epoch 3/30
625/625 [==============================] - 532s 851ms/step - loss: 0.7749 - accuracy: 0.6887 - val_loss: 0.6604 - val_accuracy: 0.6941 - lr: 0.0010
Epoch 4/30
625/625 [==============================] - 554s 886ms/step - loss: 0.6004 - accuracy: 0.7185 - val_loss: 0.6260 - val_accuracy: 0.7079 - lr: 0.0010
Epoch 5/30
625/625 [==============================] - 537s 858ms/step - loss: 0.5385 - accuracy: 0.7400 - val_loss: 0.5631 - val_accuracy: 0.7212 - lr: 0.0010
Epoch 6/30
625/625 [==============================] - 523s 837ms/step - loss: 0.5024 - accuracy: 0.7595 - val_loss: 0.5342 - val_accuracy: 0.7422 - lr: 0.0010
Epoch 7/30
625/625 [==========================

accuracy: 0.8555 - val_loss: 0.3552 - val_accuracy: 0.8456